mlflow/onnx.py

```diff
@@ -182,10 +182,27 @@ def _load_model(model_file):
 
 
 class _OnnxModelWrapper:
-    def __init__(self, path):
+    def __init__(self, path, providers=None):
```

Comment:
for this to be accessible to an end-user, the `_load_pyfunc()` function would need to be updated to support loading with a particular intended end-use environment provider configuration which would in turn change the signature of `mlflow.pyfunc.load_model()` to make this a modifiable argument. 
This is probably not what we want to do here. 
We might want to think about setting this property for an onnx flavor in a different way (during logging of the model and have this property stored in the serialized artifact?) 
WDYT @harupy @dbczumar ?

Reply:
@BenWilson2 completely understand the reservations about changing the `mlflow.pyfunc.load_model()` functionality given it's application to all model flavours.

I think your suggestion about modifying the saving metadata information on saving of the model is quite a nice suggestion as this means that only the `onnx.py` module code would need to be changed, is that right?

So under the `model_flavours` section of the `MLmodel` yaml:

```yaml
flavors:
  onnx:
    data: model.onnx
    onnx_version: 1.10.2
    providers: ['CUDAExecutionProvider', CPUExecutionProvider']
```

From my assessment, that would be facilitated through an additional argument added to the `add_flavour` method call in line 140 of `onnx.py`:

From:

`mlflow_model.add_flavor(FLAVOR_NAME, onnx_version=onnx.__version__, data=model_data_subpath)`

to:

`mlflow_model.add_flavor(FLAVOR_NAME, onnx_version=onnx.__version__, data=model_data_subpath, providers=providers)`

Could the variable `providers` be an additional optional argument for the `save_model` function, giving the user the ability to specify their own preference of providers (with a default list if not specified)?

```python
def save_model(
    onnx_model,
    path,
    conda_env=None,
    mlflow_model=None,
    signature: ModelSignature = None,
    input_example: ModelInputExample = None,
    pip_requirements=None,
    extra_pip_requirements=None,
    providers=None
):
   if not providers:
      providers = ['CUDAExecutionProvider', 'CPUExecutionProvider']

```

Reply:
@ecm200 @BenWilson2 This proposal sounds great to me!

Reply:
@dbczumar @BenWilson2 

Great. I’ll draft a change and commit and update the PR.

Only thing I need to figure is how to get the info into the ONNX private class that handles the inference session setup. 

My proposal would be that given this class gets the path to the onnx model, then the expectation is that this model will always be an MLModel type format? Therefore we can expect to have an MLModel yaml file located in the same directory? Is that a correct assumption?

If that’s a reasonable assumption then the `__init__` method of the aforementioned class could simply load the model meta data which would have the `providers` parameters set. For backwards comparability, where the MLModel file does not have a entry for `providers` then the fallback option could be to have a default list which covers a reasonable range of execution types.

Reply:
New commit af21a11 has this implemented.
See below commetns.